scheduler/core/sources.py

```diff
@@ -0,0 +1,116 @@
+# Copyright (c) 2016-2022 Association of Universities for Research in Astronomy, Inc. (AURA)
+# For license information see LICENSE or https://opensource.org/licenses/BSD-3-Clause
+
+from enum import Enum
+from abc import ABC, abstractmethod
+from typing import ClassVar, Optional, NoReturn, Tuple
+from scheduler.services.resource import OcsResourceService, FileResourceService
+from scheduler.services.environment import Env
+from io import BytesIO
+
+class Services(Enum):
+    ENV = 'env'
+    RESOURCE = 'resource'
+    CHRONICLE = 'chronicle'
+
+class Origin(ABC):
+
+
+    def __init__(self,
+                 resource: ClassVar[Optional[OcsResourceService]] = None,
+                 env: ClassVar[Optional[Env]] = None,
+                 chronicle = None, # Missing typing for lack of implementation
+                 is_loaded: ClassVar[bool] = False):
+        self.resource = resource
+        self.env = env
+        self.chronicle = chronicle
+        self.is_loaded = is_loaded
+
+    @abstractmethod
+    def load(self) -> NoReturn:
+        raise NotImplementedError('load Origin')
+
+
+    def __str__(self):
+        return self.__class__.__name__.replace('Origin','').upper()
+
+
+class OCSOrigin(Origin):
+
+    def load(self) -> 'OCSOrigin':
+        if not self.is_loaded:
+            self.resource = OcsResourceService()
+            self.env = Env()
+            # OCSOrigin.chronicle
+            self.is_loaded = True
+            return self
+        return self
+
+
+class GPPOrigin(Origin):
+    def load(self) -> NoReturn:
+        raise NotImplementedError('GPP sources are not implemented')
+
+class FileOrigin(Origin):
+    def load(self):
+        return self
+
+class Origins(Enum):
+    FILE = FileOrigin()
+    OCS = OCSOrigin()
+    GPP = GPPOrigin()
```

Comment:
As mentioned over Slack, perhaps:
```python
class Origins(Enum):
    FILE = lambda: FileOrigin()
    OCS = lambda: OCSOrigin()
    GPP = lambda: GPPOrigin()
```
to avoid creating all three when we only need one.